### PR TITLE
Refactoring `_print_name` for certain RVs and specifying `rv_type`s in their distributions

### LINKED ISSUE #6201: Create PyMC-specific RV Ops that overwrite `_print_name`
Currently, some RV's `_print_name`s contain the abbreviated version of the random variable Op rather than its full name, e.g. Normal, Cauchy, etc. When it come to displaying them in Graphviz, we resort to its class name via [`some_op.__class__.__name__`](https://github.com/pymc-devs/pymc/blob/main/pymc/model_graph.py#L165), which, to my understanding, is a bit hackish. A solution would look like the following:

```python
PyMCNormalRV(NormalRV):
    _print_name = ("Normal", ...)

pymc_normal = PyMCNormalRV()

class Normal(Continuous):
    rv_op = pymc_normal
```

In particular, with the newly added `SymbolicRV`s, this could be helpful (see WIP PR #6149).

CC @ricardoV94 @michaelosthege

### PR DESCRIPTION
**Pull request by larryshamalama**
**What is this PR about?**
This PR creates "new" PyMC `RandomVariable`s for those that have an abbreviate `_print_name` such that we don't have to rely on the RV's class name for Latex and Graphviz displays.

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## Bugfixes / New features
- ...

## Docs / Maintenance
- 
Subclass several random variables to have a more readable `_print_name`.

Closes #6201

### CODE REVIEW

File: pymc/distributions/discrete.py [hunk 1]
```diff
@@ -674,6 +681,13 @@ def logcdf(value, mu):
         return check_parameters(res, 0 <= mu, msg="mu >= 0")
 
 
+class PyMCNegativeBinomialRV(NegBinomialRV):
+    _print_name = ("NBinom", "\\operatorname{NBinom}")
```
Comment (ricardoV94) [inline pymc/distributions/discrete.py]: Why this name?

File: pymc/tests/distributions/test_shape_utils.py [hunk 2]
```diff
@@ -573,7 +574,7 @@ def test_change_rv_size_default_update():
 def test_change_specify_shape_size_univariate():
     with aesara.config.change_flags(mode=Mode("py")):
         s1, s2 = at.iscalars("s1", "s2")
-        x = at.random.normal(size=(s1, s2))
+        x = pymc_normal(size=(s1, s2))
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_shape_utils.py]: We don't need to change the object used in tests that are not related to the docstrings

File: pymc/distributions/truncated.py [hunk 3]
```diff
@@ -338,7 +341,7 @@ def truncated_logprob(op, values, *inputs, **kwargs):
     return logp
 
 
-@_truncated.register(NormalRV)
+@_truncated.register(PyMCNormalRV)
```
Comment (ricardoV94) [inline pymc/distributions/truncated.py]: This should work on the base class (I hope) and we should keep it like that.

File: pymc/distributions/multivariate.py [hunk 4]
```diff
@@ -1868,7 +1884,7 @@ def rng_fn(self, rng, mu, sigma, *covs, size=None):
         if sigma:
             cov = cov + sigma**2 * np.eye(cov.shape[0])
 
-        x = multivariate_normal.rng_fn(rng=rng, mean=mu, cov=cov, size=size)
+        x = pymc_multivariate_normal.rng_fn(rng=rng, mean=mu, cov=cov, size=size)
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]: No need to change this or above. Only thing that matters is the `rv_op` we use in the distribution classes. All other changes should be undone.

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

**Comment by twiecki** [comment]
Can't we just set `_print_name` on the imported aesara rvs?

**Comment by ricardoV94** [comment]
> Can't we just set `_print_name` on the imported aesara rvs?

That would have side-effects on the Aesara side (if it even works)

**Comment by larryshamalama** [comment]
Letting tests run again. Hopefully nothing fails.

Are there any additional tests to add in light of these new `Op`s?

File: pymc/tests/distributions/test_truncated.py [hunk 5]
```diff
@@ -9,15 +9,15 @@
...
+from pymc.distributions.continuous import Exponential, PyMCNormalRV, TruncatedNormalRV
 from pymc.distributions.shape_utils import change_dist_size
 from pymc.distributions.transforms import _default_transform
 from pymc.distributions.truncated import Truncated, TruncatedRV, _truncated
 from pymc.exceptions import TruncationError
 from pymc.tests.distributions.util import assert_moment_is_expected
 
 
-class IcdfNormalRV(NormalRV):
+class IcdfNormalRV(PyMCNormalRV):
```
Comment (larryshamalama) [inline pymc/tests/distributions/test_truncated.py]: In light of this [failing test](https://github.com/pymc-devs/pymc/actions/runs/3274707968/jobs/5388603941), this test RV needs to inherit from `PyMCNormalRV` because the normal distribution's moment is defined under `Normal` which uses the `pymc_normal` `rv_op`.

Comment (ricardoV94) [inline pymc/tests/distributions/test_truncated.py]: In that case let's use `PyMC` versions in all these test classes? Less confusing I think

Comment (larryshamalama) [reply pymc/tests/distributions/test_truncated.py]: I added a comment to clarify why we are doing this then. Let me know if you think that this is good

File: pymc/tests/distributions/test_truncated.py [hunk 6]
```diff
@@ -6,25 +6,32 @@
...
     """Normal RV that has icdf but not truncated dispatching"""
 
 
-class RejectionNormalRV(NormalRV):
+class RejectionNormalRV(PyMCNormalRV):
     """Normal RV that has neither icdf nor truncated dispatching."""
 
 
+"""
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_truncated.py]: I would also do it for the GeometricRV since we also use a PyMCGeometricRV internally. The tests work regardless but I think it's less confusing for future users.

**Comment by ricardoV94** [comment]
I am not super happy with this (even though I suggested it). It's a bit worrisome that our dispatched methods will work for the subclasses but not the original classes, when all we are doing is changing a name. 

We could do something more clever, where we still dispatch on the Aesara classes but return the PyMC ones from Distribution when those are specified. Then both objects will have logp/logcdf/moments/ etc...

That would have avoided the concerns about the tests where we create new classes.

We could use the `rv_type` property, and dispatch on that. So Normal would be

```python
class Normal(Continuous):
  rv_op = pymc_normal
  rv_type = NormalRV  # Aesara
  ...
```
And then tweak the Metaclass of Distribution to use `rv_type` for dispatching when provided explicitly, and otherwise default to the old behavior of obtaining the `rv_type` from `rv_op`. WDYT?

**Comment by larryshamalama** [comment]
> I am not super happy with this (even though I suggested it). It's a bit worrisome that our dispatched methods will work for the subclasses but not the original classes, when all we are doing is changing a name.

I understand. I also had some doubts, but at least we have a clearer picture of how the changes in this PR would affect the codebase.

> We could do something more clever, where we still dispatch on the Aesara classes but return the PyMC ones from Distribution when those are specified. Then both objects will have logp/logcdf/moments/ etc...
> 
> That would have avoided the concerns about the tests where we create new classes.
> 
> We could use the `rv_type` property, and dispatch on that. So Normal would be
> 
> ```python
> class Normal(Continuous):
>   rv_op = pymc_normal
>   rv_type = NormalRV  # Aesara
>   ...
> ```
> 
> And then tweak the Metaclass of Distribution to use `rv_type` for dispatching when provided explicitly, and otherwise default to the old behavior of obtaining the `rv_type` from `rv_op`. WDYT?

On top of my head, both solutions would work, but I'm not sure how I would dispatch the Aesara classes given `rv_op` to be the newly created ones (like `PyMCNormalRV`).

https://github.com/pymc-devs/pymc/blob/13dfeb2d72a001776b0d0408cf28302e826240cd/pymc/distributions/distribution.py#L107-L114

I believe that the dispatch is being done in these lines and perhaps there is a more clever way than checking if `__class__.__name__.startswith("PyMC")`... Is this what you meant?

Again, without trying much, I feel like using `rv_type` would be an easier solution. However, would this have any spillover complications in other areas of the codebase?

**Comment by ricardoV94** [comment]
@larryshamalama what I meant was something like:

```python
        rv_type = clsdict.setdefault("rv_type", None)

        if rv_type is None and isinstance(rv_op, RandomVariable):
            rv_type = type(rv_op)
            clsdict["rv_type"] = rv_type
```

Then when we set:
```python
class Normal(Continuous):
  rv_op = pymc_normal
  rv_type = NormalRV  # Aesara
```
`pm.Normal` will still return a `PyMCNormalRV`, but all the dispatching will be done on the original baseclass. Dispatching works for subclasses but not parent classes (which makes sense).

**Comment by larryshamalama** [comment]
Okay, I can get to it

File: pymc/distributions/logprob.py [hunk 7]
```diff
@@ -145,7 +145,8 @@ def joint_logp(
 
     The input `var` determines which log-likelihood graph is used and
     `rv_value` is that graph's input parameter.  For example, if `var` is
-    the output of a ``NormalRV`` ``Op``, then the output is a graph of the
+    the output of a ``NormalRV`` ``Op``, which has been recently adapted to
```
Comment (ricardoV94) [inline pymc/distributions/logprob.py]: We can leave this unchanged

File: pymc/tests/distributions/test_truncated.py [hunk 8]
```diff
@@ -25,6 +25,13 @@ class RejectionNormalRV(NormalRV):
...
+Note: IcdfNormalRV and RejectionNormalRV inherit from PyMCNormalRV rather
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_truncated.py]: Can remove comment

File: pymc/tests/test_aesaraf.py [hunk 9]
```diff
@@ -405,7 +406,8 @@ def test_rvs_to_value_vars_unvalued_rv():
     res_y = res.owner.inputs[1]
     # Graph should have be cloned, and therefore y and res_y should have different ids
     assert res_y is not y
-    assert res_y.owner.op == at.random.normal
+    assert res_y.owner.op == pymc_normal
+    assert isinstance(res_y.owner.op, NormalRV)
```
Comment (ricardoV94) [inline pymc/tests/test_aesaraf.py:408]: Second assert is enough since this is not a pm.Distribution test

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
@larryshamalama Can you include the default changelist from the PR template? We use those for the auto-generated release notes

**Comment by larryshamalama** [comment]
Done! Thanks for the suggestion, I will start doing this more often

**Comment by ricardoV94** [comment]
Thanks @larryshamalama, now the `print_name_` for SymbolicRVs should be more useful
